jekyll/_posts/2021-03-12-new-Nim-channels.md

```diff
@@ -0,0 +1,206 @@
+```
+
+**app.nim**
+
+The main thread prepares tasks by reading `todo_urls.json`, and then it sends JSON data to a channel. The crawl thread does the actual work - it receives URL data from the channel and downloads the contents using the `httpclient` module.
+
+```nim
+import std/channels
+import std/[httpclient, isolation, json]
+var ch = newChannel[JsonNode]() # we need to send JsonNode
+proc download(client: HttpClient, url: string) =
+  let response = client.get(url)
+  echo "content: ", response.body[0 .. 20] # prints the results
+proc crawl =
+  var client = newHttpClient() # the crawler
+  defer: client.close()
+  let data = ch.recv() # the JSON data
+  if data != nil:
+    for url in data["url"]:
+      download(client, url.getStr)
+proc prepareTasks(fileWithUrls: string): seq[Isolated[JsonNode]] =
+  result = @[]
+  for line in lines(fileWithUrls):
+    result.add isolate(parseJson(line)) # parse JSON file
+proc spawnCrawlers =
+  var tasks = prepareTasks("todo_urls.json")
+  for t in mitems tasks: # we need a mutable view of the items
+    ch.send move t
+var thr: Thread[void]
+createThread(thr, crawl) # create crawl thread
+spawnCrawlers()
+joinThread(thr)
+```
+
+First, we import `std/channels`.
+
+Then we can create a channel using `newChannel`, which returns a `Channel[T]`. It uses `mpmc` internally, which stands for "multiple producer, multiple consumer". The `elements` parameter is used to specify whether a channel is buffered or not. For an unbuffered channel, the sender and the receiver block until the other side is ready. Sending data to a buffered channel blocks only when the buffer is full. Receiving data from a buffered channel blocks when the buffer is empty.
+
+`newChannel` is a generic proc - you can specify the types of the data you want to send or receive.
+
+```nim
+var chan1 = newChannel[int]()
+# or
+var chan2 = newChannel[string](elements = 1) # unbuffered channel
+# or
+var chan3 = newChannel[seq[string]](elements = 30) # buffered channel
+```
+
+The `send` proc takes data that we want to send to the channel. The passed data is moved around, not copied. Because `chan.send(isolate(data))` is very common to use, `template send[T](c: var Chan[T]; src: T) = chan.send(isolate(src))` is provided for convenience. For example, you can use `chan.send("Hello World")` instead of `chan.send(isolate("Hello World!"))`.
+
+There are two useful procs for a receiver: `recv` and `tryRecv`. `recv` blocks until something is sent to the channel. In contrast, `tryRecv` doesn't block - if no message exists in the channel, it just fails and returns `false`. We can write a while loop to call `tryRecv`and handle a message when available.
+
+### It is safe and convenient
+
+The Nim compiler rejects the program below at compile time. It says that `expression cannot be isolated: s`. This is because `s` is a `ref object` - it may be modified somewhere and is not unique, so the variable cannot be isolated.
```

Comment:
```suggestion
The Nim compiler rejects the program below. It says that `expression cannot be isolated: s`. This is because `s` is a `ref object` - it may be modified somewhere and is not unique, so the variable cannot be isolated.
```
The "at compile time" is redundant, since the Nim *compiler* rejects it, alternatively, one could write "The program below is rejected at compile time".